### DISCUSSION TITLE
Describe query's result

### DISCUSSION THREAD

**Discussion by ahka01** [comunica/comunica#868] [Q&A] [answered]
Hi everybody,

I have the following question related to using `DESCRIBE` query:

how could I make the query describe also the objects that are blank or unnamed node?

Considering the following simple tripels:

```ttl
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix ex: <http://example.com> .

ex:a a ex:b;
    ex:c ex:d;
    ex:e [
        a ex:f;
        ex:a ex:m
    ];
.
```
The results to describe the resource `ex:a` looks like:
```n3
<http://example.coma> a <http://example.comb>;
    <http://example.comc> <http://example.comd>;
    <http://example.come> _:bc_0_n3-02.
```

What I wish it to have is to return also what does `_:bc_0_n3-02` contains. Is there a way to achieve that?

PS: Fuseki returns the description of objects of triples if it is from kind blank nodes as well.

Thanks a lot in advance for answering and for the great job of comunica :)

Best regards

**Comment by rubensworks** [comment, marked as answer]
The problem with `DESCRIBE` is that it's not very strictly defined in the SPARQL spec. So executing the same `DESCRIBE` query in different engines may produce different results.

My recommendation would be to use `CONSTRUCT` queries instead, as you can specifically define the template of triples you want to produce.

Using a `CONSTRUCT WHERE`, and `FILTER` to check for blank nodes, it should be possible to drill down on blank node entities. (I don't have a query at hand for this though)

**Reply by ahka01** [reply]
Thanks a lot for the hint

**Comment by ahka01** [comment]
applying the following `DESCRIBE` query:

```sparql
DESCRIBE ?o {
    ?s ?p ?o .
        FILTER(isBlank(?o)) 
}
```

on these dummy triples:

```ttl
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix ex: <http://example.com/> .

ex:a a ex:b;
    ex:c ex:d;
    ex:e [
        a ex:f;
        ex:v ex:m ;
    ]
.
```
I'm getting the following results:

```
subject                      predicate                                                                          object                                                                                         
----------------------------------------------------------------------------------------------------------------------------
bc_0_n3-00            http://www.w3.org/1999/02/22-rdf-syntax-ns#type     http://example.com/f                                                                                 
bc_0_n3-01             http://example.com/v                                                       http://example.com/m  
```
I'm wondering why the the subjects here have different labels? They shouldn't be the same?

**Reply by rubensworks** [reply]
@jacoscaz You did some work around this topic recently, right? Was this intentional behaviour for construct queries? I forgot what the findings of our latest discussions were...

**Reply by jacoscaz** [reply]
Where is `http://example.com/v` coming from? It's not in the triples above if I am not mistaken.

**Reply by ahka01** [reply]
I edited the comment again by adding the triples, that I'm referring to. They are indeed different from the one in the main question. Sorry for that!

**Reply by jacoscaz** [reply]
@ahka01 thanks for updating! 

I don't have the time to look further into this today but this looks like a bug having something to do with blank node correlation. @rubensworks I have indeed done some work on this topic as a fix to https://github.com/beautifulinteractions/node-quadstore/issues/134 . [This comment](https://github.com/beautifulinteractions/node-quadstore/issues/134#issuecomment-808955790) lists the relevant upstream commits.

**Reply by jacoscaz** [reply]
[This comment](https://github.com/comunica/comunica/issues/773#issuecomment-822495504) might also be relevant here (or it might not, approach with caution as with everything related to blank nodes):

> The spec says that "blank node labels are scoped to the template for each solution". Since there are 4 solutions here, there are 4 template instantiations, each with their own (non-clashing) bnode.

Perhaps there's something that's triggering a new template instantiation per each quad even in this case?

**Reply by rubensworks** [reply]
> Perhaps there's something that's triggering a new template instantiation per each quad even in this case?

Yeah, it might be related to that.

@ahka01 If you would, you could open a bug report and/or PR with a fix. In any case, I think using CONSTRUCT queries might be a better solution going forward.
